### DISCUSSION TITLE
Elsa 3.0 Send Email activity not storing the 'To' address in the workflow definition

### DISCUSSION THREAD

**Discussion by markjones-WK** [elsa-workflows/elsa-core#4595] [Q&A]
I am using the latest Elsa 3.0 and when configuring a simple workflow definition using the send email activity, I cannot get the To: field (or any recipient fields) to save/persist when I save and publish the workflow definition.

IN other words, I am trying to do a very simple workflow definition where there is an http endpoint and a send email activity.  When I enter an email address into the To field and save the workflow definition and then refresh the browser (or navigate elsewhere and then back to workflow definitions), the email I entered into the To field is empty and hence when I run the hit the endpoint, I am seeing a 'No recipients have been specified.' error in the logs of the docker container.

Is this a bug or what am I doing wrong?

**Comment by markjones-WK** [comment]
FYI... I was able to get it working, but I had to configure the workflow definition differently.  Instead of typing the To: email address into the To: field directly, I used a variable and assigned a default value and added the variable to the To: field instead and that seemed to allow it to save/persist.

I would still like to know if this is a bug or by design?

**Reply by sfmskywalker** [reply]
What you described is definitely a bug. You should be able to enter a literal value with it getting lost.
Can you tell me more about what version of latest Elsa 3.0? Is it a preview release from feedz.io, or from NuGet?
Also, are you using the Stencil implementation of the designer (deprecated), or the latest Blazor implementation of Elsa Studio?

When I try using the latest source code on my machine of both elsa-core:v3 and elsa-studio:main, I cannot reproduce the behavior you described.

**Reply by markjones-WK** [reply]
I can reproduce it using the latest docker hub all-in-one image, 'elsaworkflows/elsa-v3:latest'.  I believe I was also able to reproduce it from the source code from the v3 branch, but cant be sure at the moment.

**Reply by sfmskywalker** [reply]
I see. I tried the Docker image just now, and I was not able to reproduce the issue. Please make sure you're using the latest Image version by first deleting the image; at least that is what I have to do always as to not start an older, cached version.

This is what I see after refreshing the page:

<img width="1447" alt="image" src="https://github.com/elsa-workflows/elsa-core/assets/938393/8a0d97c2-22a0-4c7a-ac51-1ea457e85d54">

The values are still there. Also, please make sure there's no outdated JS files in the browser's cache.

**Reply by markjones-WK** [reply]
Interesting.  I will try again adn will delete the browser cache.  I had already tried an Incognito browser mode.  I also believe I deleted the image as well, but will try again.  Will let you know a bit later.

Thanks!

**Reply by markjones-WK** [reply]
Also fyi, after I had entered the email address into the To: field, I never did see the 'blue circle' around the email address.  Is that supposed to render like that after leaving the field in the editor or after saving?

**Reply by markjones-WK** [reply]
Ok.  My apologies.  I re-did it all and it works as expected now.  Thanks!

**Reply by sfmskywalker** [reply]
No worries, I’m happy it works!

**Reply by markjones-WK** [reply]
I was able to reproduce this again and now know the steps to reproduce.

When in the send email activity properties, I enter an email address into the To: field (or any recipient field for that matter) and then click out of the To: field or directly click save.  The workflow is saved and when browser is refreshed, there is no value saved in the To: field.

Now, if I do it again and this time after I enter the email address into the To: field, I hit the Tab key (which validated and puts the blue circle around it) and then click save, the workflow definition is saved and the value is persisted.

So, it appears that I must click the tab button before clicking elsewhere or saving the workflow definition in order for the values to be persisted/saved.

Can you see this behavior?  If so, is this by design or bug?

Thanks!

**Reply by sfmskywalker** [reply]
I see! Definitely not by my design, but it would appear this is default behavior for the [MudChips component from MudBlazor](https://codebeam-mudextensions.pages.dev/mudchipfield). On their demo page, you see the same behavior - you have to press the delimiter key (which in their demo is the space character). Perhaps this behavior is configurable, but I didn't find anything to that effect (yet).

**Reply by sfmskywalker** [reply]
Actually, upon further inspection, I do see an `OnBlur` event - so, if the component exposes a method to "press delimiter", then that might fix the issue you discovered. I will have a closer look.

**Reply by sfmskywalker** [reply]
Although I managed to get the text value and insert it as a badge (aka chip), the MudChips component doesn't seem to expose any way to clear out the text input. So now I am not sure what is worse: to leave the text input entry + seeing the chip being added (but at least the contents are saved), or leaving the default behavior (but with the potential loss of the entry that was about to be added) 🤔

**Reply by aaronscribner** [reply]
@sfmskywalker 
I work with @markjones-WK , FYI
Is it possible to define the 2-way model binding on the component?

https://stackoverflow.com/questions/57932850/how-to-make-two-way-binding-on-blazor-component

**Reply by markjones-WK** [reply]
Another option might be to make the recipient fields like the From: field where is does not use the 'chips' and make it a semi-colon delimited string for multiple email addresses? I can also enter two email addresses in the To: field with a semi-solon delimiter and then hit the tab key and both show up in the same 'chip', which I also don't think is expected?

![image](https://github.com/elsa-workflows/elsa-core/assets/126109777/9d26b777-8bc7-4f67-bfba-ede4ac5998dc)

**Reply by sfmskywalker** [reply]
That's certainly an option, I'll think about it. As for the semi-colon thing, ideally you would be able to press either tab, space or semi-colon to create a new chip, but unfortunately, the Chips component only allows for one delimiter. But I wouldn't be surprised if thy add an option to allow multiple delimiters at some point.
